ratelimits/bucket.go

```diff
@@ -1,52 +1,253 @@
+	bucketId, err := newIPAddressBucketId(NewRegistrationsPerIPAddress, ip)
+	if err != nil {
+		return Transaction{}, err
+	}
+	return newTransaction(bucketId, cost), nil
+}
+
+// RegistrationsPerIPv6RangeTransaction returns a Transaction for the
+// NewRegistrationsPerIPv6Range limit for the /48 IPv6 range which contains the
+// provided IPv6 address.
+func RegistrationsPerIPv6RangeTransaction(ip net.IP, cost int64) (Transaction, error) {
+	bucketId, err := newIPv6RangeCIDRBucketId(NewRegistrationsPerIPv6Range, ip)
+	if err != nil {
+		return Transaction{}, err
+	}
+	return newTransaction(bucketId, cost), nil
+}
+
+// OrdersPerAccountTransaction returns a Transaction for the NewOrdersPerAccount
+// limit for the provided ACME registration Id.
+func OrdersPerAccountTransaction(regId, cost int64) (Transaction, error) {
+	bucketId, err := newRegIdBucketId(NewOrdersPerAccount, regId)
+	if err != nil {
+		return Transaction{}, err
+	}
+	return newTransaction(bucketId, cost), nil
+}
+
+// FailedAuthorizationsPerAccountCheckOnlyTransaction returns a 0 cost
+// check-only Transaction for the provided ACME registration Id for the
+// FailedAuthorizationsPerAccount limit.
+func FailedAuthorizationsPerAccountCheckOnlyTransaction(regId int64) (Transaction, error) {
+	bucketId, err := newRegIdBucketId(FailedAuthorizationsPerAccount, regId)
+	if err != nil {
+		return Transaction{}, err
+	}
+	return newCheckOnlyTransaction(bucketId, 0), nil
+}
+
+// FailedAuthorizationsPerAccountTransaction returns a Transaction for the
+// FailedAuthorizationsPerAccount limit for the provided ACME registration Id.
+func FailedAuthorizationsPerAccountTransaction(regId, cost int64) (Transaction, error) {
+	bucketId, err := newRegIdBucketId(FailedAuthorizationsPerAccount, regId)
+	if err != nil {
+		return Transaction{}, err
+	}
+	return newTransaction(bucketId, cost), nil
+}
+
+// CertificatesPerDomainTransactions returns a slice of Transactions for the for
+// the provided order domain names. An error is returned if any of the order
+// domain names are invalid. When a CertificatesPerDomainPerAccount override is
+// configured, two types of Transactions are returned:
+//   - CertificatesPerDomain Transaction(s), which will NOT be denied if the
+//     bucket lacks the capacity to satisfy the cost, and
+//   - a CertificatesPerDomainPerAccount Transaction, which will be denied if
+//     the bucket lacks the capacity to satisfy combined cost of each
+//     CertificatesPerDomain Transaction(s).
+//
+// When a CertificatesPerDomainPerAccount override is NOT configured, only
+// CertificatesPerDomain Transactions, which will be denied if the bucket lacks
+// the required capacity, are returned.
+func CertificatesPerDomainTransactions(limiter *Limiter, regId int64, orderDomains []string, cost int64) ([]Transaction, error) {
+	certsPerDomainPerAccountId, err := newRegIdBucketId(CertificatesPerDomainPerAccount, regId)
```

Comment:
I think this doesn't match the semantics of the current certificates per domain rate limit with an account override. As far as I can tell, this creates one bucket per account and charges it for N items with each order. But instead there should be one bucket per (account, eTLD+1) pair, and each bucket should be charged for the number of items that fall into that bucket. E.g.:

new-order for regID 1234, domains `example.com`, `www.example.com`, `example.net`

bucket (1234, `example.com`) gets charged for 2
bucket (1234, `example.net`) gets charged for 1

Reply:
Summary of verdict reached offline:

- Attach the limit to the Transaction before Spend/Refund time.
- Add new bucketKey format `enum:regId:domain`
- When an override exists for CertificatesPerDomain `enum:regId:domain`:
  - Return a spend-only Transaction for each per domain bucket.
  - Return a check-and-spend Transaction for each per account per domain bucket, using the new `enum:regId:domain` bucketKey format.
- Optional: Move knowledge of limits out of the `Limiter` to a new exported struct.
- Optional: Attach `Transaction` constructors to the aforementioned new exported struct.